### PR TITLE
Fixes #13349: accept Search on subparts of an identifier, not only on subidentifiers of an identifier

### LINKED ISSUE #13349: Search: string search tries to interpret strings as references.
#### Description of the problem

This is another regression of the new search command over ssrsearch: one can no longer search numbers as substrings of lemma names. 

```coq
Require List.
Search "eq" inside List.
Fail Search "1" inside List. 
Require ssrsearch.
Search "1" in List.
```
The Fail command fails with: `Unable to interpret "1" as a reference` which is strange given that this is supposed to search for substrings in lemma names. 

#### Coq Version

8.12.0

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** fix

Fixes / closes #13349

- [X] Added / updated test-suite
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Comment by herbelin** [comment]
Actually, this creates an ambiguity. If I do `Search "'mod'".`, it might mean a subpart `'mod'` of an identifier (even if unrealistically wanted), or the infix notation `mod`. I guess the solution is to force the infix notation to be written instead `"_ mod _"`, or `(_ mod _)`... No objections to that?

**Comment by Zimmi48** [comment]
> Actually, this creates an ambiguity. If I do `Search "'mod'".`, it might mean a subpart `'mod'` of an identifier (even if unrealistically wanted), or the infix notation `mod`. I guess the solution is to force the infix notation to be written instead `"mod"%nat`, `"_ mod _"`, or `(_ mod _)`... No objections to that?

I didn't realize that `'mod'` is a valid subpart of an identifier (I'm not sure it's such a good idea that it is the case BTW).

This is slightly annoying because the documentation contains advice to use single quotes for disambiguation.

I'm afraid changing this would be a bit much for 8.12.1 then. So either, we make a special case for single quotes, or we have to fix the doc and postpone to 8.13, or we do this change and at the same time disallow single quotes in the middle of an identifier (only allow them at the end), but this also means postponing to 8.13 (which is not far off anyway).

File: doc/changelog/07-commands-and-options/13351-master+search-all-subparts-ident.rst [hunk 1]
```diff
@@ -0,0 +1,6 @@
+- **Fixed:**
+  :cmd:`Search` supports filtering on parts of identifiers which are
+       not proper identifiers themselves, such as :n:`"1"`
```
Comment (Zimmi48) [inline doc/changelog/07-commands-and-options/13351-master+search-all-subparts-ident.rst:3]:
```suggestion
  not proper identifiers themselves, such as :n:`"1"`
```

**Comment by herbelin** [comment]
> I didn't realize that 'mod' is a valid subpart of an identifier (I'm not sure it's such a good idea that it is the case BTW).

> This is slightly annoying because the documentation contains advice to use single quotes for disambiguation.

> I'm afraid changing this would be a bit much for 8.12.1 then. So either, we make a special case for single quotes, or we have to fix the doc and postpone to 8.13

I guess that in general `'` is used at the end of identifier, so, I don't think it would be a problem to keep the convention that strings surrounded by simple quotes are considered as notations, i.e. to make a special case for single quotes.

**Review by Zimmi48 — APPROVED** [review]

**Comment by Zimmi48** [comment]
I'll merge this PR soon if there are no more comments.

**Comment by Zimmi48** [comment]
@coqbot: merge now

**Pull request merged**